vllm/multimodal/base.py

```diff
@@ -43,6 +43,13 @@ def try_concat(
         *,
         device: torch.types.Device,
     ) -> BatchedTensors:
+        # may be list rather than tensors
+        if isinstance(tensors[0], list):
+            new_tensors = []
+            for new_tensor in tensors:
+                for new_t in new_tensor:
+                    new_tensors.append(new_t.to(device))
+            return new_tensors
```

Comment:
It seems that you're flattening the nested list of tensors here into a (non-nested) list of tensors. Wouldn't that break the convention that the number of tensors should equal the batch size?

Reply:
First, I wanna ask that is there any batched data input support for vllm? As I only see the `input_ids` of model input is in 1-D shape. So I think batch size indicates multi-image inputs here?
Second, When it comes to vision language models, here the `tensors`, which is one of the input parameters, might be in shape `[B, _, 3, H, W]`. The second dimension will be `1` for in most cases, while in some image-slicing cases, like LLava-Next or our model, the second dimension might be greater than 1. In LLava-Next, the shape of sliced images will be the same, so you can use `torch.cat` . But in MiniCPM-V the sliced image might be in different shapes, so I need to use this way as concatenating.

Reply:
We currently only support single image inputs. The batch size corresponds to the number of examples.

Reply:
Just noticed your comment that this 2D case can also occur for single image inputs. In that case, you should update the batcher to output 2D nested list (rather than flattening them into a 1D list), otherwise it will cause conflicts when we need to also support multiple images input.

Reply:
Could I ask more questions about this? Actually I'm a little confused with it. The `input_ids` in `forward` of these models is in 1-D shape, so why the input images have a "Batch" dimension?

Reply:
The `input_ids` etc. have been formatted to work with KV cache so they have been flattened. However, multi-modal inputs are passed directly to the model so they retain the original batch dimensions.

Reply:
> Just use prompts parameter to input `List[str]` to llm

Actually this is batch input. Each `str` represents a single input.

Reply:
After offline discussion, I have suggested this function to handle arbitrary nested lists of tensors:

```py
# Similar to original, but handles empty lists just in case
# collate_fn_map is required for pytorch collate (see below)
def _try_concat(tensors: List[torch.Tensor], *, device: torch.types.Device, collate_fn_map: object = None) -> Union[torch.Tensor, List[torch.Tensor]]:
    if len(tensors) == 0:
        return tensors

    unbatched_shape = tensors[0].shape[1:]

    for tensor in tensors:
        if tensor.shape[1:] != unbatched_shape:
            return [
                tensor.squeeze(0).to(device=device) for tensor in tensors
            ]

    return torch.cat(tensors, dim=0).to(device=device)

# Leverage the existing function in pytorch
from functools import partial
from torch.utils.data._utils.collate import collate

NestedTensors = Union[List["NestedTensors"], List[torch.Tensor]]

def _try_concat_nested(tensors: NestedTensors, *, device: torch.types.Device) -> Union[torch.Tensor, NestedTensors]:
    return collate(tensors, collate_fn_map={torch.Tensor: partial(_try_concat, device=device)})  # type: ignore
```

Hope this helps!